### PR TITLE
feat(webscene): ArcGIS WebScene loader

### PR DESCRIPTION
**Pull request by dryabinin-actionengine**
SPEC - https://developers.arcgis.com/web-scene-specification/objects/webscene/

### CODE REVIEW

File: modules/i3s/src/i3s-webscene-loader.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import type {LoaderWithParser, LoaderOptions, LoaderContext} from '@loaders.gl/loader-utils';
+import type {I3SWebScene} from './types';
+
+import {parseI3SWebscene} from './lib/parsers/parse-i3s-webscene';
+
+// __VERSION__ is injected by babel-plugin-version-inline
+// @ts-ignore TS2304: Cannot find name '__VERSION__'.
+
```
Comment (belom88) [inline modules/i3s/src/i3s-webscene-loader.ts:8]: Remove empty line as comment on 7th line is for 9th line.

File: modules/i3s/src/lib/parsers/parse-i3s-webscene.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ */
+function parseOperationalLayers(
+  layersList: I3SWebSceneOperationalLayer[]
+): I3SWebSceneOperationalLayer[] {
+  let layers: I3SWebSceneOperationalLayer[] = [];
+
+  for (let index = 0; index < layersList.length; index++) {
+    const layer = layersList[index];
+    // Add support only for scene service layer for now.
+    // TODO Add ArcGISFeatureLayer support.
```
Comment (belom88) [inline modules/i3s/src/lib/parsers/parse-i3s-webscene.ts:34]: This is a requirement. I'm not sure we need some special todo here. May be just comment that @loaders.gl supports only 3D scene layers.

**Review by belom88 — CHANGES_REQUESTED** [review]

File: modules/i3s/src/lib/parsers/parse-i3s-webscene.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import type {I3SWebScene, I3SWebSceneOperationalLayer} from '../../types';
+
+const ARCGIS_SCENE_SERVER_LAYER_TYPE = 'ArcGISSceneServiceLayer';
```
Comment (ibgreen) [inline modules/i3s/src/lib/parsers/parse-i3s-webscene.ts:3]: Is this an ARCGIS specific loader? If so should the name reflect that? Similar to Cesium ION loader?

File: modules/i3s/src/types.ts [hunk 4]
```diff
@@ -771,3 +771,141 @@ export type FieldInfo = {
...
+type I3SWebSceneOperationalLayerDefinition = {
+  elevationInfo: I3SWebSceneOperationalLayerElevationInfo;
+  drawingInfo: I3SWebSceneOperationalLayerDrawingInfo;
+};
+
+type I3SWebSceneOperationalLayerElevationInfo = {
+  mode: string;
+};
+
+type I3SWebSceneOperationalLayerDrawingInfo = {
```
Comment (ibgreen) [inline modules/i3s/src/types.ts:820]: Can we get a JSDoc on every type? Links to spec?

**Review by ibgreen — APPROVED** [review]
- tests?
- docs?
- release notes?

File: modules/i3s/test/i3s-web-scene-layer.spec.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
+import test from 'tape-promise/tape';
```
[the remark this answers is not in the document]
Comment (dryabinin-actionengine) [reply modules/i3s/test/i3s-web-scene-layer.spec.js]: Renamed

File: modules/i3s/src/types.ts [hunk 6]
```diff
@@ -771,3 +771,141 @@ export type FieldInfo = {
...
+
+export type I3SWebSceneLayer = {
+  operationalLayers: I3SWebSceneOperationalLayer[];
+  baseMap: I3SWebSceneBaseMap;
+  ground: I3SWEbSceneGround;
+  heightModelInfo: HeightModelInfo;
+  version: string;
+  authoringApp: string;
+  authoringAppVersion: string;
+  // TODO Add presentation type. This data allows to add multiple viewes for particular web scene.
```
[the remark this answers is not in the document]
Comment (dryabinin-actionengine) [reply modules/i3s/src/types.ts]: Done

File: modules/i3s/src/types.ts [hunk 7]
```diff
@@ -771,3 +771,141 @@ export type FieldInfo = {
   isEditable: boolean;
   label: string;
 };
+
+export type I3SWebScene = {
+  header: I3SWebSceneLayer;
```
[the remark this answers is not in the document]
Comment (dryabinin-actionengine) [reply modules/i3s/src/types.ts]: Done

File: modules/i3s/src/types.ts [hunk 8]
```diff
@@ -771,3 +771,141 @@ export type FieldInfo = {
...
+  showLegend?: boolean;
+  layers?: I3SWebSceneOperationalLayer[];
+};
+
+type I3SWebSceneOperationalLayerDefinition = {
+  elevationInfo: I3SWebSceneOperationalLayerElevationInfo;
+  drawingInfo: I3SWebSceneOperationalLayerDrawingInfo;
+};
+
+type I3SWebSceneOperationalLayerElevationInfo = {
```
[the remark this answers is not in the document]
Comment (dryabinin-actionengine) [reply modules/i3s/src/types.ts]: Renamed all items and saved energy =)

File: modules/i3s/src/i3s-webscene-loader.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  extensions: ['json'],
+  options: {}
+};
+
+async function parse(
+  data: ArrayBuffer,
+  options?: LoaderOptions,
+  context?: LoaderContext
+): Promise<I3SWebScene> {
+  if (!context?.url) {
```
[the remark this answers is not in the document]
Comment (dryabinin-actionengine) [reply modules/i3s/src/i3s-webscene-loader.ts]: Removed

File: modules/i3s/src/i3s-webscene-loader.ts (hunk 1 above)
Comment (dryabinin-actionengine) [reply modules/i3s/src/i3s-webscene-loader.ts]: Done

File: modules/i3s/src/lib/parsers/parse-i3s-webscene.ts (hunk 2 above)
Comment (dryabinin-actionengine) [reply modules/i3s/src/lib/parsers/parse-i3s-webscene.ts]: Change logic

File: modules/i3s/src/lib/parsers/parse-i3s-webscene.ts (hunk 3 above)
Comment (dryabinin-actionengine) [reply modules/i3s/src/lib/parsers/parse-i3s-webscene.ts]: Renamed it to `ArcGisWebSceneLoader`.

File: modules/i3s/src/types.ts (hunk 4 above)
Comment (dryabinin-actionengine) [reply modules/i3s/src/types.ts]: Done

**Comment by dryabinin-actionengine** [comment]
> * tests?
> * docs?
> * release notes?

1. Done
2. Done
3. Will be done in separate PR.

File: modules/i3s/src/lib/parsers/parse-arcgis-webscene.ts [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+const SUPPORTED_LAYERS_TYPES = [
+  ARCGIS_SCENE_SERVER_LAYER_TYPE,
+  INTEGRATED_MESH_LAYER,
+  BUILDING_SCENE_LAYER
+];
+
+/**
+ * Parses ArcGIS WebScene
+ * @param data
+ * @param options
```
Comment (belom88) [inline modules/i3s/src/lib/parsers/parse-arcgis-webscene.ts:20]: there is no options

File: modules/i3s/src/lib/parsers/parse-arcgis-webscene.ts [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  return {
+    header: layer0,
+    layers: parseOperationalLayers(operationalLayers)
+  };
+}
+
+/**
+ * Recursively parses WebScene operational layers.
+ * @param sublayers
+ * @param url
```
Comment (belom88) [inline modules/i3s/src/lib/parsers/parse-arcgis-webscene.ts:36]: there is no url

File: modules/i3s/src/lib/parsers/parse-arcgis-webscene.ts [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+  return {
+    header: layer0,
+    layers: parseOperationalLayers(operationalLayers)
+  };
+}
+
+/**
+ * Recursively parses WebScene operational layers.
+ * @param sublayers
```
Comment (belom88) [inline modules/i3s/src/lib/parsers/parse-arcgis-webscene.ts:35]: name doesn't match to the function's argument

File: modules/i3s/src/types.ts [hunk 13]
```diff
@@ -772,84 +772,141 @@ export type FieldInfo = {
...
+   * An object used to specify the spatial reference of the given geometry.
+   */
   spatialReference: SpatialReference;
+  /**
+   * Type: String
+   * Valid values:
+   * global
+   * local
+   */
   viewingMode: string;
```
Comment (belom88) [inline modules/i3s/src/types.ts:839]: `viewingMode: 'global' | 'local'`

File: modules/i3s/src/types.ts [hunk 14]
```diff
@@ -772,84 +772,141 @@ export type FieldInfo = {
...
+  /**
+   * An object used to specify the spatial reference of the given geometry.
+   */
   spatialReference: SpatialReference;
+  /**
+   * Type: String
+   * Valid values:
+   * global
+   * local
+   */
```
Comment (belom88) [inline modules/i3s/src/types.ts:838]: the comment content can be declared in the type

File: modules/i3s/test/index.js [hunk 15]
```diff
@@ -6,3 +6,4 @@ import './i3s-node-page-loader.spec';
 import './i3s-attribute-loader.spec';
 import './i3s-content-loader.spec';
 import './i3s-building-scene-layer-loader.spec';
+import './arcgis-web-scene-loader.spec';
```
Comment (belom88) [inline modules/i3s/test/index.js:9]:
```suggestion
import './arcgis-webscene-loader.spec';
```

File: modules/i3s/test/arcgis-web-scene-loader.spec.js [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+            }
+          ]
+        },
+        type: 'simple'
+      }
+    }
+  }
+};
+
+const ARCGIS_WEB_SCENE_URL = '@loaders.gl/i3s/test/data/ArcGisWebScene/arcGisWebScene.json';
```
Comment (belom88) [inline modules/i3s/test/arcgis-web-scene-loader.spec.js:42]:
```suggestion
const ARCGIS_WEB_SCENE_URL = '@loaders.gl/i3s/test/data/arcgis-webscenes/arcgis-webscene.json';
```

**Review by belom88 — APPROVED** [review]

File: modules/i3s/src/lib/parsers/parse-arcgis-webscene.ts (hunk 12 above)
Comment (dryabinin-actionengine) [reply modules/i3s/src/lib/parsers/parse-arcgis-webscene.ts]: Renamed

File: modules/i3s/src/types.ts (hunk 14 above)
Comment (dryabinin-actionengine) [reply modules/i3s/src/types.ts]: Done

File: modules/i3s/src/types.ts (hunk 13 above)
Comment (dryabinin-actionengine) [reply modules/i3s/src/types.ts]: Done

File: modules/i3s/test/arcgis-web-scene-loader.spec.js (hunk 16 above)
Comment (dryabinin-actionengine) [reply modules/i3s/test/arcgis-web-scene-loader.spec.js]: Done

**Pull request merged**